File: Golang/strategy/duck.go
package strategy

import "fmt"

type Duck struct {
    FlyBehavior FlyBehavior
    QuackBehavior QuackBehavior
}

func (d Duck) PerformFly() {
    d.FlyBehavior.Fly()
}

func (d Duck) PerformQuack() {
    d.QuackBehavior.Quack()
}

func (d Duck) Swim() {
    fmt.Println("All ducks float!")
}

func (d *Duck) SetFlyBehavior(fb FlyBehavior) {
    d.FlyBehavior = fb
}

func (d *Duck) SetQuackBehavior(qb QuackBehavior) {
    d.QuackBehavior = qb
}

type MallardDuck struct {
    Duck
}

func NewMallardDuck() *MallardDuck {
    mallard := &MallardDuck{}
    mallard.FlyBehavior = FlyWithWings{}
    mallard.QuackBehavior = Quack{}
    return mallard
}

func (md MallardDuck) Display() {
    fmt.Println("I'm a Mallard Duck")
}

File: Python/strategy/duck.py
from abc import ABC, abstractmethod
from fly_behavior import FlyBehavior
from quack_behavior import QuackBehavior


class Duck(ABC):
    flyBehavior: FlyBehavior
    quackBehavior: QuackBehavior

    @abstractmethod
    def display(self) -> None:
        raise NotImplementedError

    def setFlyBehavior(self, fb: FlyBehavior) -> None:
        self.flyBehavior = fb

    def setQuackBehavior(self, qb: QuackBehavior) -> None:
        self.quackBehavior = qb

    def performFly(self) -> None:
        self.flyBehavior.fly()

    def performQuick(self) -> None:
        self.quackBehavior.quack()

    def swim(self) -> None:
        print("All ducks float!")

